VirtoCommerce.Platform.Data/Assets/AssetEntryService.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using CacheManager.Core;
+using VirtoCommerce.Platform.Core.Assets;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Data.Common;
+using VirtoCommerce.Platform.Data.Infrastructure;
+using VirtoCommerce.Platform.Data.Model;
+using VirtoCommerce.Platform.Data.Repositories;
+
+namespace VirtoCommerce.Platform.Data.Assets
+{
+    public class AssetEntryService : ServiceBase, IAssetEntryService, IAssetEntrySearchService
+    {
+        private readonly Func<IPlatformRepository> _platformRepository;
+        private readonly IBlobUrlResolver _blobUrlResolver;
+        private readonly ICacheManager<object> _cacheManager;
+
+        public AssetEntryService(Func<IPlatformRepository> repositoryFactory, IBlobUrlResolver blobUrlResolver, ICacheManager<object> cacheManager)
+        {
+            _platformRepository = repositoryFactory;
+            _blobUrlResolver = blobUrlResolver;
+            _cacheManager = cacheManager;
+        }
+
+        public AssetEntrySearchResult SearchAssetEntries(AssetEntrySearchCriteria criteria)
+        {
+            var cacheKey = $"Search:{criteria?.GetHashCode()}";
+            return _cacheManager.Get(cacheKey, nameof(AssetEntry), () =>
+            {
+                criteria = criteria ?? new AssetEntrySearchCriteria();
+
+                using (var repository = _platformRepository())
+                {
+                    var query = repository.AssetEntries;
+
+                    if (!string.IsNullOrEmpty(criteria.SearchPhrase))
+                    {
+                        query = query.Where(x => x.Name.Contains(criteria.SearchPhrase) || x.RelativeUrl.Contains(criteria.SearchPhrase));
+                    }
+
+                    if (!string.IsNullOrEmpty(criteria.Language))
+                    {
+                        query = query.Where(x => x.LanguageCode == criteria.Language);
+                    }
+
+                    if (!string.IsNullOrEmpty(criteria.Group))
+                    {
+                        query = query.Where(x => x.Group == criteria.Group);
+                    }
+
+                    if (!criteria.Tenants.IsNullOrEmpty())
+                    {
+                        var tenants = criteria.Tenants.Where(x => !string.IsNullOrEmpty(x.TenantId) && !string.IsNullOrEmpty(x.TenantType)).ToArray();
```

Comment:
This logic can be moved to Tenant type `IsValid` or `IsWellFormed` properties

Reply:
Need to remove these  and replace to follow code
```
var tenantsStrings = tenants.select(x=> x.ToString())
query = query.Where(x=> tenantsStrings.Contains(x.TenantId + "_" + x.tenantType)
```

Reply:
any particular reason for
query = query.Where(x=> tenantsStrings.Contains(x.TenantId + "_" + x.tenantType)
?

Reply:
Do you really think what is your solution more clear than mine?

Reply:
somehow I was expecting to learn about performance, execution plans and indices from you in here. But you're dead simple:)